### PR TITLE
ParserUtil: Update header comments on Optional usage #557

### LINKED ISSUE #557: ParserUtil: Update header comments on Optional usage
ParserUtil#parseX(Optional<String>) methods currently take in an Optional<String>. This violates the usage of Java Optional.

However, ParserUtil is currently being used like a middle-man as it takes in an Optional from the caller, and it returns a parsed Optional to be used by the caller. It seems redundant to first unwrap the initial Optional, just to have it re-wrap back into an Optional after being parsed by ParserUtil.

Let's add header comments in ParserUtil to make it clear to future developers why ParserUtil takes in Optional<String> even though it violates Java's convention.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #557

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/1] [ParserUtil: Update header comments on Optional usage](https://github.com/se-edu/addressbook-level4/pull/558/commits/50e5ded88572e8a5d27573e34835097c21534216)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/56e874d0e7b4cab2b87d573782df3b34f9bafc3c...50e5ded88572e8a5d27573e34835097c21534216))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/558/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/1] [ParserUtil: Update header comments on Optional usage](https://github.com/se-edu/addressbook-level4/pull/558/commits/7871039e0e048e14fe2c33cce5c7a354835a2e7b)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/56e874d0e7b4cab2b87d573782df3b34f9bafc3c...7871039e0e048e14fe2c33cce5c7a354835a2e7b)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/4229d6f22bf8ed5356cfa29e1edf8ba74e669330~1...4229d6f22bf8ed5356cfa29e1edf8ba74e669330))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/558/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -17,7 +17,12 @@
...
 
 /**
- * Contains utility methods used for parsing strings in the various *Parser classes
+ * Contains utility methods used for parsing strings in the various *Parser classes.
+ * {@code ParserUtil} contains methods that take in {@code Optional} as parameters. However, it goes against Java's
+ * convention (see https://stackoverflow.com/a/39005452) as {@code Optional} should only be used a return type.
+ * Going against the convention here is acceptable as {@code ParserUtil} is currently being used like a middle-man.
+ * It seems redundant to first unwrap the initial {@code Optional}, just to have it re-wrap back into an
+ * {@code Optional} after being parsed by {@code ParserUtil}.
  */
```
Comment (damithc) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: Should we add something like `See header comment regarding the use of Optional parameters` in the methods in concern too?

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/1] [ParserUtil: Update header comments on Optional usage](https://github.com/se-edu/addressbook-level4/pull/558/commits/bd7eaaa7cd1e5e50b017e10864a53c0113bdb46b)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/56e874d0e7b4cab2b87d573782df3b34f9bafc3c...bd7eaaa7cd1e5e50b017e10864a53c0113bdb46b)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/ce7af7dfa8b6ef3392d7d5a9c1215e8df0dce255~1...ce7af7dfa8b6ef3392d7d5a9c1215e8df0dce255))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/558/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 1 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: Noted, made the changes.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -17,7 +17,12 @@
 import seedu.address.model.tag.Tag;
 
 /**
- * Contains utility methods used for parsing strings in the various *Parser classes
+ * Contains utility methods used for parsing strings in the various *Parser classes.
+ * {@code ParserUtil} contains methods that take in {@code Optional} as parameters. However, it goes against Java's
+ * convention (see https://stackoverflow.com/a/39005452) as {@code Optional} should only be used a return type.
+ * Going against the convention here is acceptable as {@code ParserUtil} is currently being used like a middle-man.
+ * It seems redundant to first unwrap the initial {@code Optional}, just to have it re-wrap back into an
```
Comment (limmlingg) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: `It is redundant to unwrap the initial {@code Optional}, then re-wrap it into a {@code Optional} after being parsed by {@code ParserUtil}.`

Same for the commit message.

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/1] [ParserUtil: Update header comments on Optional usage](https://github.com/se-edu/addressbook-level4/pull/558/commits/ebf8dbd67b92d1b8cebc0679b4ec64a8a2cc4daf)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/52ec52609b8a17da202dd3fa8f9813b7820b5d3f...ebf8dbd67b92d1b8cebc0679b4ec64a8a2cc4daf)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/ec44c5dd59d681952d2d58537a7ffd38706c95dd~1...ec44c5dd59d681952d2d58537a7ffd38706c95dd))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/558/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@damithc for your final review :>

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -39,6 +44,7 @@ public static Index parseIndex(String oneBasedIndex) throws IllegalValueExceptio
 
     /**
      * Parses a {@code Optional<String> name} into an {@code Optional<Name>} if {@code name} is present.
+     * See header comment regarding the use of Optional parameters.
```
Comment (damithc) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: `header comment of this class`?
`{@code Optional}`?

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -17,7 +17,12 @@
 import seedu.address.model.tag.Tag;
 
 /**
- * Contains utility methods used for parsing strings in the various *Parser classes
+ * Contains utility methods used for parsing strings in the various *Parser classes.
+ * {@code ParserUtil} contains methods that take in {@code Optional} as parameters. However, it goes against Java's
+ * convention (see https://stackoverflow.com/a/39005452) as {@code Optional} should only be used a return type.
+ * Going against the convention here is acceptable as {@code ParserUtil} is currently being used like a middle-man.
```
Comment (damithc) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: A little more details can be useful to the reader.
~~Going against the convention here is acceptable as {@code ParserUtil} is currently being used like a middle-man.~~
`Justification: The methods in concern receive {@code Optional} return values from other methods as parameters and return  {@code Optional} values based on whether the parameters were present. Therefore, it is redundant to unwrap the initial {@code Optional} before passing to {@code ParserUtil} as a parameter and then re-wrap it into an {@code Optional} return value inside {@code PrarserUtil} methods.`

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/1] [ParserUtil: Update header comments on Optional usage](https://github.com/se-edu/addressbook-level4/pull/558/commits/da3e0dca18ce55ad01fce4bc632269d9083f89a2)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/52ec52609b8a17da202dd3fa8f9813b7820b5d3f...da3e0dca18ce55ad01fce4bc632269d9083f89a2)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/190c8aac1487795d7ac04fca659219991bf005f8~1...190c8aac1487795d7ac04fca659219991bf005f8))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/558/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@yamgent should be mergeable now :)

**Comment by Zhiyuan-Amos** [comment]
> Copypaste-fu very good.

Yup. Seems like I didn't do this PR cos I just copied and pasted

> Can merge after that.

no need last call la :P

**Pull request merged**